core/src/main/java/org/elasticsearch/common/limit/LimitingStreamOutput.java

```diff
@@ -0,0 +1,87 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.elasticsearch.common.limit;
+
+import org.elasticsearch.Version;
+import org.elasticsearch.common.io.stream.StreamOutput;
+
+import java.io.IOException;
+
+/**
+ * A stream wrapper that limits the number of bytes that can be written. The reserved amount is reduced when the stream is closed or reset.
+ */
+public class LimitingStreamOutput extends StreamOutput {
+    private final StreamOutput delegate;
+    private final Limiter limiter;
+    private long bytesWritten;
+
+    public LimitingStreamOutput(StreamOutput delegate, Limiter limiter) {
+        this.delegate = delegate;
+        this.limiter = limiter;
+    }
+
+    @Override
+    public void writeByte(byte b) throws IOException {
+        addByteChecked();
+        delegate.writeByte(b);
+    }
+
+    @Override
+    public void writeBytes(byte[] b, int offset, int length) throws IOException {
+        addBytesChecked(length);
+        delegate.writeBytes(b, offset, length);
+    }
+
+    @Override
+    public void flush() throws IOException {
+        delegate.flush();
+    }
+
+    @Override
+    public void close() throws IOException {
+        addBytesChecked(-bytesWritten);
+        delegate.close();
+    }
+
+    @Override
+    public void reset() throws IOException {
+        addBytesChecked(-bytesWritten);
```

Comment:
Looking at the implementation:

```java
    @Override
    public void reset() {
        // shrink list of pages
        if (bytes.size() > BigArrays.PAGE_SIZE_IN_BYTES) {
            bytes = bigarrays.resize(bytes, BigArrays.PAGE_SIZE_IN_BYTES);
        }

        // go back to start
        count = 0;
    }
```

It doesn't necessarily "free" bytes, just resizes them back to a 16384 byte array if more than 16384 bytes have been written already, so I'm not sure what to do here